File: server/src/main/java/org/tbbtalent/server/model/db/partner/Partner.java
/*
 * Copyright (c) 2022 Talent Beyond Boundaries.
 *
 * This program is free software: you can redistribute it and/or modify it under
 * the terms of the GNU Affero General Public License as published by the Free
 * Software Foundation, either version 3 of the License, or any later version.
 *
 * This program is distributed in the hope that it will be useful, but WITHOUT
 * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
 * FITNESS FOR A PARTICULAR PURPOSE. See the GNU Affero General Public License
 * for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program. If not, see https://www.gnu.org/licenses/.
 */

package org.tbbtalent.server.model.db.partner;

import org.springframework.lang.NonNull;
import org.springframework.lang.Nullable;
import org.tbbtalent.server.model.db.Status;
import org.tbbtalent.server.model.db.User;

/**
 * A partner is an organization that works to implement displaced talent mobility.
 * <p/>
 * All Talent Catalog users are associated with just one partner at a time.
 *
 * @author John Cameron
 */
public interface Partner extends SourcePartner {

    /**
     * Abbreviated name of partner, if any.
     * @return Abbreviated name of partner - eg "TBB", or null if no abbreviation.
     */
    @Nullable
    String getAbbreviation();
    void setAbbreviation(@Nullable String s);

    /**
     * True if this partner is the default source partner - associated with candidates who are
     * not clearly associated with any other source partner.
     * <p/>
     * Only one source partner at any given time can be the default.
     * @return True if this is the default source partner
     */
    boolean isDefaultSourcePartner();
    void setDefaultSourcePartner(boolean b);

    boolean isJobCreator();
    void setJobCreator(boolean b);

    boolean isSourcePartner();
    void setSourcePartner(boolean b);

    /**
     * True if this partner is the default job creator.
     * <p/>
     * Only one partner at any given time can be the default.
     * @return True if this is the default job creator
     */
    boolean isDefaultJobCreator();
    void setDefaultJobCreator(boolean b);

    @Nullable
    Long getContextJobId();
    void setContextJobId(@Nullable Long contextJobId);

    @Nullable
    User getDefaultContact();
    void setDefaultContact(@Nullable User defaultContact);

    /**
     * Unique id identifying this partner
     * @return partner id
     */
    Long getId();
    void setId(Long id);

    /**
     * Gets the contact associated with {@link #getContextJobId()} if it is not null, otherwise
     * returns the default partner contact, {@link #getDefaultContact()}
     * @return Contact user
     */
    @Nullable
    User getJobContact();

    /**
     * Partner's logo.
     * <p/>
     * Optional - if none is specified this call returns null, and the calling code may default
     * to using a default logo.
     * @return Link to partner's logo.
     */
    @Nullable
    String getLogo();
    void setLogo(@Nullable String s);

    /**
     * Name of partner.
     * @return Name of partner - eg "Talent Beyond Boundaries"
     */
    @NonNull
    String getName();
    void setName(@NonNull String s);

    /**
     * Email used to notify partner
     * @return Partner email
     */
    @Nullable
    String getNotificationEmail();
    void setNotificationEmail(@Nullable String notificationEmail);

    /**
     * Salesforce ID (extracted from {@link #getSflink()}).
     */
    @Nullable
    String getSfId();

    /**
     * Url link to corresponding Salesforce Account record, if one exists.
     */
    @Nullable
    String getSflink();
    void setSflink(@Nullable String sflink);

    /**
     * Status of partner. Partners may be inactivated or deleted.
     * @return Status of partner
     */
    @NonNull
    Status getStatus();
    void setStatus(@NonNull Status s);

    /**
     * Url of partner's website (optional)
     * @return Website url, null if none - eg "https://talentbeyondboundaries.org/"
     */
    @Nullable
    String getWebsiteUrl();
    void setWebsiteUrl(@Nullable String s);

}

File: ui/admin-portal/src/app/model/partner.ts
/*
 * Copyright (c) 2024 Talent Catalog.
 *
 * This program is free software: you can redistribute it and/or modify it under
 * the terms of the GNU Affero General Public License as published by the Free
 * Software Foundation, either version 3 of the License, or any later version.
 *
 * This program is distributed in the hope that it will be useful, but WITHOUT
 * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
 * FITNESS FOR A PARTICULAR PURPOSE. See the GNU Affero General Public License
 * for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program. If not, see https://www.gnu.org/licenses/.
 */
import {Country} from "./country";
import {User} from "./user";
import {JobChat} from "./chat";

/*
  MODEL - extending interfaces, update request extend object attributes, mapping enums

  - Avoid duplicating lists of attributes by using inheritance (ie extends).
  - Link update request to partner attributes (ie don't duplicate attributes)
*/

export interface PartnerSimpleAttributes {
  abbreviation: string;
  autoAssignable: boolean;
  defaultPartnerRef: boolean;
  employerSflink?: string;
  jobContact?: User;
  jobCreator: boolean;
  logo: string;
  name: string;
  notificationEmail: string;
  publicApiAccess: boolean;
  publicApiAuthorities: string[];
  publicApiKey?: string;
  registrationLandingPage: string;
  sflink: string;
  sourcePartner: boolean;

  //Note that status is a Java Status enum on the server, but it maps to the string value of
  //enum by the JSON processing (Javascript does not know about enums).
  //However we do create a matching Typescript Status enum, and use that to process this
  //value in a type safe way. See other MODEL's for details.
  status: string;
  websiteUrl: string;
  redirectPartner?: ShortPartner;
}

export enum PublicApiAuthority {
  READ_CANDIDATE_DATA = "READ_CANDIDATE_DATA",
  SUBMIT_JOB_MATCHES = "SUBMIT_JOB_MATCHES",
  OFFER_CANDIDATE_SERVICES = "OFFER_CANDIDATE_SERVICES",
  REGISTER_CANDIDATES = "REGISTER_CANDIDATES",
  ADMIN = "ADMIN"
}

export interface Employer {
  id: number;
  name: string;
  description: string;
  hasHiredInternationally: boolean;
  sfId: string;
  website: string;
}

export interface ShortPartner {
  id: number;
  name: string;
  abbreviation: string;
  websiteUrl: string;
}

export interface Partner extends PartnerSimpleAttributes {
  id: number;
  defaultContact?: User;
  defaultJobCreator: boolean;
  defaultSourcePartner: boolean;
  employer?: Employer;
  sourceCountries: Country[];

  //Temporary place to store job chat associated with partner - will depend on current job context.
  //The underscore naming is a convention to distinguish this field from those that are
  //uploaded from the server - not set according to context on the browser.
  _jobChat?: JobChat;
  // DPA fields for versioned acceptance management
  acceptedDataProcessingAgreementId?: string;
  acceptedDataProcessingAgreementDate?: string;
  firstDpaSeenDate?: string;
}

export function sourceCountriesAsString(partner: Partner): string {
  let s = '';
  const countries: Country[] = partner.sourceCountries;
  if (countries != null) {
    s = countries.map(c => c.name).join(",");
  }
  return s;
}

export interface UpdatePartnerRequest extends PartnerSimpleAttributes {
  defaultContactId?: number,
  sourceCountryIds: number[];
  redirectPartnerId?: number;
}

export interface UpdatePartnerJobContactRequest {
  jobId: number;
  userId: number;
}

